File: shared/core/layout_variant.rb
# frozen_string_literal: true

module JsonUIShared
  # Helpers for responsive layout variant files (`home@regular.json`).
  #
  # A variant is a sibling of a base screen layout whose stem is
  # `<base>@<sizeClass>`; it replaces the WHOLE tree when the runtime
  # size class matches (no partial merge). The v1 file-suffix vocabulary
  # is compact / medium / regular only — landscape and combined forms
  # stay inline-`responsive` territory. Constraint validation lives in
  # `jui build` (_check_variant_constraints); the tools only need to
  # recognize variants, exclude them from independent screen generation
  # and attach them to their base.
  #
  # Mirrors jui_tools/jui_cli/core/layout_variant.py — keep in sync.
  #
  # Usage:
  #   LayoutVariant.variant?('home@regular')        # => true
  #   LayoutVariant.split('home@regular')           # => ['home', 'regular']
  #   LayoutVariant.split('home')                   # => ['home', nil]
  #   LayoutVariant.variants_for('/x/home.json')    # => {'regular' => '/x/home@regular.json'}
  #
  module LayoutVariant
    VALID_VARIANT_CLASSES = %w[compact medium regular].freeze

    # Split a layout file stem into [base, size_class]; size_class is nil
    # for base files. Splits on the LAST '@'.
    def self.split(stem)
      return [stem, nil] unless stem.include?('@')

      idx = stem.rindex('@')
      [stem[0...idx], stem[(idx + 1)..]]
    end

    # True when the stem (or file path) carries an '@' variant suffix.
    # Invalid suffixes still count — exclusion must not depend on the
    # gate having accepted the file.
    def self.variant?(path_or_stem)
      File.basename(path_or_stem.to_s, '.json').include?('@')
    end

    # Existing variant files for a base layout path, keyed by size class
    # (valid classes only, resolution order compact → medium → regular).
    def self.variants_for(base_path)
      dir = File.dirname(base_path)
      stem = File.basename(base_path, '.json')
      VALID_VARIANT_CLASSES.each_with_object({}) do |cls, found|
        candidate = File.join(dir, "#{stem}@#{cls}.json")
        found[cls] = candidate if File.exist?(candidate)
      end
    end
  end
end

File: jui_tools/jui_cli/core/layout_variant.py
"""Variant-file helpers for responsive layout variants (``home@regular.json``).

A layout variant is a sibling file of a base screen layout whose stem is
``<base>@<sizeClass>``. The variant replaces the WHOLE tree when the
runtime size class matches; there is no partial merge (structure sharing
stays with include/style). v1 accepts the single-size-class vocabulary
(compact / medium / regular) only — landscape and combined forms remain
inline-``responsive`` territory.

Mirrors ``shared/core/layout_variant.rb`` — keep the two in sync.
"""

from __future__ import annotations

# v1 file-suffix vocabulary. The inline `responsive` attribute keeps its
# full 7-value vocabulary (see responsive_resolver.rb) — only these three
# are legal as a file suffix.
VALID_VARIANT_CLASSES: tuple[str, ...] = ("compact", "medium", "regular")

# Recognized inline-responsive vocabulary that v1 rejects as a file suffix.
INLINE_ONLY_CLASSES: tuple[str, ...] = (
    "landscape",
    "compact-landscape",
    "medium-landscape",
    "regular-landscape",
)


def split_variant(stem: str) -> tuple[str, str | None]:
    """Split a layout file stem into ``(base, size_class)``.

    ``home@regular`` → ``("home", "regular")``; ``home`` → ``("home", None)``.
    Splits on the LAST ``@`` so a nested ``a@b@c`` yields base ``a@b`` —
    the constraint gate rejects that base separately.
    """
    if "@" not in stem:
        return stem, None
    base, _, cls = stem.rpartition("@")
    return base, cls


def is_variant(stem: str) -> bool:
    """True when the stem carries an ``@`` variant suffix (any suffix —
    invalid suffixes still make the file a variant for exclusion purposes,
    the gate reports them)."""
    return "@" in stem


def variant_struct_stem(base: str, cls: str) -> str:
    """The synthetic snake stem reserved for the generated variant view
    (``home`` + ``regular`` → ``home_regular_variant``). A real layout with
    this stem would collide with the generated type name — the gate
    rejects that."""
    return f"{base}_{cls}_variant"
